File: 25_is_subsequence.py
"""
Task 25: Check if one string is a subsequence of another.
Reads two lines: s and t, prints YES if s is subsequence of t.
"""
import sys

def is_subsequence(s: str, t: str) -> bool:
    i = 0
    for ch in t:
        if i < len(s) and s[i] == ch:
            i += 1
    return i == len(s)

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    t = sys.stdin.readline().rstrip('\n')
    print('YES' if is_subsequence(s,t) else 'NO')

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 25_is_subsequence.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 25: Check if one string is a subsequence of another.
Reads two lines: s and t, prints YES if s is subsequence of t.
*/

int main()
{
    string s, t;
    if (!getline(cin, s))
        return 0;
    if (!getline(cin, t))
        return 0;
    int i = 0, j = 0;
    while (i < s.size() && j < t.size())
    {
        if (s[i] == t[j])
            ++i;
        ++j;
    }
    cout << (i == s.size() ? "YES" : "NO") << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
